### PR TITLE
8268857: Merge VM_PrintJNI and VM_PrintThreads and remove the unused field 'is_deadlock' of DeadlockCycle

### PR DESCRIPTION
**Pull request by D-D-H**
Hi,

Could I have a review of this change that merges three vm operations(VM_PrintThreads, VM_PrintJNI, VM_FindDeadlocks) in thread_dump and signal_thread_entry.

`jstack` is a very common command, even in the production environment.

In addition to reduce the cost of entering safepoint, I think this patch also could ensure the consistency of the results of VM_PrintThreads and VM_FindDeadlocks.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8268857](https://bugs.openjdk.java.net/browse/JDK-8268857): Merge VM_PrintJNI and VM_PrintThreads and remove the unused field 'is_deadlock' of DeadlockCycle

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4504/head:pull/4504` \
`$ git checkout pull/4504`

Update a local copy of the PR: \
`$ git checkout pull/4504` \
`$ git pull https://git.openjdk.java.net/jdk pull/4504/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4504`

View PR using the GUI difftool: \
`$ git pr show -t 4504`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4504.diff">https://git.openjdk.java.net/jdk/pull/4504.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
Sorry but I do not agree with this change. I understand your rationale but you have overlooked that by combining these three safepoint operations you have created a much longer safepoint pause which may lock out other more important safepoint operations, and you may also be holding the HeapLock across this extended safepoint.

The operations that you have coalesced are not considered time critical but can themselves be very time consuming e.g deadlock detection.

David

**Comment by D-D-H** [comment]
> Sorry but I do not agree with this change. I understand your rationale but you have overlooked that by combining these three safepoint operations you have created a much longer safepoint pause which may lock out other more important safepoint operations, and you may also be holding the HeapLock across this extended safepoint.
> 
> The operations that you have coalesced are not considered time critical but can themselves be very time consuming e.g deadlock detection.
> 
> David

Hi David,

Thanks for the comment.

I still think it's redundant to enter three safepoint, which may lead to inconsistent results between 'PrintThreads' and 'FindDeadlocks'. But this patch does cause the problem you said

I will reopen it once I have a better solution.

Thanks,
Denghui

**Comment by tstuefe** [comment]
It would be nice to have this fixed though. Maybe adding a guaranteed timeout to these operations would help.

**Comment by dholmes-ora** [comment]
So the PR is open again but I see no substantive changes that address my concerns. I don't see how there can be any fix here when the desired goal is one safepoint and the problem is that one safepoint may be far too long. And I don't see any way to force a timeout in a VM operation, or anything it calls.

I realize there can be a discrepancy between the sets of threads involved, but I think that is just the trade-off we have to make for this diagnostic tool.

**Comment by dholmes-ora** [comment]
Sorry still not a fan even with the hacked in deadline support for find_deadlocks. It is too arbitrary - you have no idea what constitutes a reasonable deadline, nor by how much you may overshoot that deadline.

David

**Comment by D-D-H** [comment]
> Sorry still not a fan even with the hacked in deadline support for find_deadlocks. It is too arbitrary - you have no idea what constitutes a reasonable deadline, nor by how much you may overshoot that deadline.
> 
> David

Hi David,

Thanks for the comment.

Yes, there is no good way to determine a suitable deadline.

I intend to temporarily give up the idea of combining VM_FindDeadlocks into VM_ExtendedPrintThreads.

But I think we can at least merge `VM_PrintJNI` and `VM_PrintThreads`, because the implementation of `VM_PrintJNI` is very fast. What do you think?

Denghui

File: src/hotspot/share/runtime/os.cpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/os.cpp:402]: Please comment what the false/true arguments mean e.g.

... false /* no extended info \*/, true /* print JNI handle info \*/);

File: src/hotspot/share/runtime/vmOperation.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vmOperation.hpp:112]: This change is not needed now.

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vmOperations.hpp:146]: Please call this _print_jni_handle_info and update parameter names accordingly.

Though it seems we never pass false here and so always print the JNI handle info, in which case we can get rid of the field and the parameter.

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vmOperations.hpp:150]: This no-arg constructor is unused now.

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (dholmes-ora) [inline src/hotspot/share/runtime/vmOperations.hpp:151]: As far as I can see you don't rely on the default parameter value so it is not needed.

File: src/hotspot/share/services/attachListener.cpp
Comment (dholmes-ora) [inline src/hotspot/share/services/attachListener.cpp:187]: Please comment what the true argument means.

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (dholmes-ora) [inline src/hotspot/share/services/diagnosticCommand.cpp:538]: Please comment what the true argument means.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
This looks much better. A few minor code changes requested - and possibly a big simplification.

Thanks,
David

File: src/hotspot/share/services/diagnosticCommand.cpp
Comment (D-D-H) [reply src/hotspot/share/services/diagnosticCommand.cpp]: fixed

File: src/hotspot/share/services/attachListener.cpp
Comment (D-D-H) [reply src/hotspot/share/services/attachListener.cpp]: fixed

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (D-D-H) [reply src/hotspot/share/runtime/vmOperations.hpp]: fixed

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (D-D-H) [reply src/hotspot/share/runtime/vmOperations.hpp]: the no-arg  constructor is used by JVM_DumpAllStacks now.

File: src/hotspot/share/runtime/vmOperations.hpp
Comment (D-D-H) [reply src/hotspot/share/runtime/vmOperations.hpp]: fixed.

File: src/hotspot/share/runtime/vmOperation.hpp
Comment (D-D-H) [reply src/hotspot/share/runtime/vmOperation.hpp]: fixed.

File: src/hotspot/share/runtime/os.cpp
Comment (D-D-H) [reply src/hotspot/share/runtime/os.cpp]: fixed.

**Review by dholmes-ora — APPROVED** [review]
Nothing further from me.

Thanks,
David

**Comment by D-D-H** [comment]
@dholmes-ora 
Thank you!

@tstuefe Could you help review this change?

**Comment by D-D-H** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [serviceability-dev](mailto:serviceability-dev@mail.openjdk.java.net):_
> 
> Hi,
> 
> Please note that hotspot changes require two reviews before integrating
> (unless a trivial change).
> 
> David
> 
> On 22/06/2021 6:31 pm, Denghui Dong wrote:

Sorry about this,  I will be more aware next time.

Denghui.

**Comment by dcubed-ojdk** [comment]
This PR should also have waiting for a review from the Serviceability team.

**Review by sspitsyn — COMMENTED** [review]
Despite the fact it is after the integration it looks good to me. :)
Thanks,
Serguei

**Pull request closed without merging**
